packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/CopyFromRealmTests.kt

```diff
@@ -0,0 +1,755 @@
+                        // Order is not guaranteed in the set when round-tripped through Core.
+                        // Also HashSets on JVM are rather annoying when it comes to byte arrays.
+                        // ByteArray equals/hashcode only considers the memory address, and not
+                        // the full content, so when copying byte arrays, the JVM does not consider
+                        // them equals. So for this test, use `any` instead of `contains`.
+                        if (copiedValue == null) {
+                            assertTrue(set.contains(copiedValue))
+                        } else {
+                            assertTrue(
+                                set.any {
+                                    (it as ByteArray).contentEquals(copiedValue)
+                                },
+                                "${prop.name} failed: $copiedValue"
+                            )
+                        }
+                    }
+                } else {
+                    val copiedSet = accessor.get(unmanagedCopy) as Set<Any?>
+                    assertEquals(set.size, copiedSet.size)
+                    copiedSet.forEach { copiedValue ->
+                        // Order is not guaranteed in the set when round-tripped through Core.
+                        assertTrue(set.contains(copiedValue), "${prop.name} failed: $copiedValue")
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun objectSet() {
+        val sample = Sample().apply {
+            objectSetField = (1..5).map { i ->
+                Sample().apply { stringField = i.toString() }
+            }.toRealmSet()
+        }
+
+        val insertedObj = realm.writeBlocking {
+            copyToRealm(sample)
+        }
+        val unmanagedObj: Sample = insertedObj.copyFromRealm()
+
+        // Close Realm to ensure data is decoupled from Realm
+        realm.close()
+
+        assertNotSame(insertedObj, unmanagedObj)
+        assertNotNull(unmanagedObj.objectSetField)
+        val copiedSet: RealmSet<Sample> = unmanagedObj.objectSetField
+        assertEquals(5, copiedSet.size)
+        copiedSet.forEach { copiedObject ->
+            assertEquals(
+                1,
+                sample.objectSetField.filter {
+                    copiedObject.stringField == it.stringField
+                }.size
+            )
+        }
+    }
+
+    @Test
+    fun realmResults() {
+        realm.writeBlocking {
+            copyToRealm(Sample().apply { stringField = "sample" })
+        }
+
+        val results = realm.query<Sample>().find()
+        assertEquals(1, results.size)
+
+        val unmanagedCopy: List<Sample> = results.copyFromRealm()
+
+        // Close Realm to ensure data is decoupled from Realm
+        realm.close()
+
+        assertEquals(1, unmanagedCopy.size)
+        assertEquals("sample", unmanagedCopy.first().stringField)
+    }
+
+    @Test
+    fun invalidObject_throws() {
```

Comment:
I would separate the test for delete vs. close objects to make it simpler to reason about

Reply:
Or maybe group all the _deleted_ tests together and all the _closed_ tests together ... but I guess it is just a matter of which axises that feels more natural to you 🤪

Reply:
Also lacking tests for `copyFromRealm` on closed results.